### PR TITLE
Flesh out paxos impl

### PR DESCRIPTION
**Pull request by bIgBV**

### CODE REVIEW

File: src/consensus/mod.rs [hunk 1]
```diff
@@ -7,16 +15,28 @@ use paxos::Paxos;
...
     my_addr: Endpoint,
     size: usize,
     paxos: Paxos<'a, C>,
+    decided: AtomicBool,
+    /// Channel used to communicate with upstream about decision
+    decision_tx: oneshot::Sender<Vec<Endpoint>>,
+    /// Channel the paxos instance will use to communicate with us about decision
+    // paxos_rx: oneshot::Receiver<Vec<Endpoint>>,
+    scheduled_paxos: Option<Box<dyn Future<Output = Result<()>>>>,
+    config_id: usize,
```
Comment (LucioFranco) [inline src/consensus/mod.rs]: you should use `common::ConfigId`

File: src/consensus/mod.rs [hunk 2]
```diff
@@ -25,22 +45,38 @@ impl<'a, C, B> FastPaxos<'a, C, B> {
...
     }
 
-    pub async fn propose(&self, proposal: Vec<Endpoint>) {
-        // TODO: register proposal with regular paxos instance
-        unimplemented!();
+    pub async fn propose(&self, proposal: Vec<Endpoint>) -> Result<()> {
+        let mut paxos_delay = Delay::new(Instant::now() + self.get_random_delay()).fuse();
+
+        // a hacky way to schedule a task after the delay
+        async {
```
Comment (LucioFranco) [inline src/consensus/mod.rs:79]: not sure I understand this hack, isnt it roughly the same thing?

Comment (bIgBV) [reply src/consensus/mod.rs:79]: I wanted to have a way to schedule the classic round task in the background after a delay. I didn't know if there was a better way to do this, and it felt kinda hacky.

File: src/consensus/mod.rs [hunk 3]
```diff
@@ -35,10 +36,17 @@ pub struct FastPaxos<'a, C, B> {
     /// Channel the paxos instance will use to communicate with us about decision
     // paxos_rx: oneshot::Receiver<Vec<Endpoint>>,
     paxos: Option<Paxos<'a, C>>,
+    /// Latest configuration we accepted.
     config_id: ConfigId,
+    votes_received: HashSet<Endpoint>, // should be a bitset?
+    votes_per_proposal: HashMap<Vec<Endpoint>, AtomicUsize>,
```
Comment (LucioFranco) [inline src/consensus/mod.rs]: I think we can drop this `AtomicUsize` for a `usize`.

File: src/consensus/mod.rs [hunk 4]
```diff
@@ -47,60 +55,106 @@ impl<'a, C, B> FastPaxos<'a, C, B> {
...
         async {
             paxos_delay.await;
             self.start_classic_round().await;
-        }.await;
+        }
+            .await;
+
+        let (tx, rx) = oneshot::channel();
+
+        let kind = proto::RequestKind::Consensus(proto::Consensus::FastRoundPhase2bMessage(
```
Comment (LucioFranco) [inline src/consensus/mod.rs]: I think it'd be nice to make this into a fn on `Request`.

File: src/consensus/mod.rs [hunk 5]
```diff
@@ -47,60 +55,106 @@ impl<'a, C, B> FastPaxos<'a, C, B> {
...
+    async fn handle_fast_round<'b>(
+        &'b mut self,
+        request: &'b proto::FastRoundPhase2bMessage,
+    ) -> crate::Result<Vec<Endpoint>> {
+        if request.config_id != self.config_id {
+            return Err(Error::new_unexpected_request(None));
+        }
+
+        if self.votes_received.contains(&request.sender) {
+            return Err(Error::new_unexpected_request(None));
```
Comment (LucioFranco) [inline src/consensus/mod.rs]: is this the correct error type?

File: src/consensus/mod.rs [hunk 6]
```diff
@@ -47,60 +55,106 @@ impl<'a, C, B> FastPaxos<'a, C, B> {
...
+        }
+
+        if (self.decided.load(Ordering::SeqCst)) {
```
Comment (LucioFranco) [inline src/consensus/mod.rs]: same here I think we can make this a regular `bool`.

**Review by LucioFranco — COMMENTED** [review]
looking good

File: src/consensus/mod.rs (hunk 3 above)
Comment (bIgBV) [reply src/consensus/mod.rs]: Yeah probably, I figured a good first step would be to copy the logic, and then think about modifying it. Though I wonder why they chose to use an `AtomicUsize` for this when it doesn't look like the hasmap is being accessed from anywhere else.

File: src/consensus/mod.rs (hunk 6 above)
Comment (bIgBV) [reply src/consensus/mod.rs]: I don' know if that's a good idea, as there could be different futures reading from the `decided` variable.

File: src/consensus/mod.rs [hunk 7]
```diff
@@ -1,58 +1,191 @@
...
+        Broadcast, Client, Request, Response,
+    },
 };
 
+const BASE_DELAY: u64 = 1000;
+
+/// Represents an instance of the Fast Paxos consensus protocol.
+///
+/// This protocol has a fast path as compared to paxos, when a majority set agrees on the same
+/// membershipset, each node individually accepts the value without the involvement of a leadero
```
Comment (LucioFranco) [inline src/consensus/mod.rs]: nit make leadero on the newline

File: src/consensus/mod.rs [hunk 8]
```diff
@@ -1,58 +1,191 @@
...
+/// r/coordinator.
+///
+/// If a qorum is not forumed, then it falls back to an instance of regular Paxos, which is
```
Comment (LucioFranco) [inline src/consensus/mod.rs]:
```suggestion
/// If a quorum is not formed, then it falls back to an instance of regular Paxos, which is
```

File: src/consensus/mod.rs (hunk 2 above)
Comment (LucioFranco) [reply src/consensus/mod.rs:79]: isnt this just the same as the two awaits one after each other?

File: src/consensus/mod.rs [hunk 9]
```diff
@@ -1,58 +1,191 @@
...
+                    Err(Error::new_unexpected_request(None))
+                }
+            }
             _ => Err(Error::new_unexpected_request(None)),
         }
     }
 
-    async fn handle_fast_round(&self, request: Request) -> crate::Result<Response> {
-        unimplemented!()
+    async fn handle_fast_round<'b>(
```
Comment (LucioFranco) [inline src/consensus/mod.rs]: can we check if we can drop these lifetimes on the latest nightly? I think that pr got merged

File: src/consensus/mod.rs [hunk 10]
```diff
@@ -1,58 +1,191 @@
...
+    async fn start_classic_round(&mut self) -> Result<()> {
+        if !self.decided.load(Ordering::SeqCst) {
+            if let Some(paxos) = &mut self.paxos {
+                paxos.start_round().await?;
+            }
+        }
+        Ok(())
+    }
+
+    fn get_random_delay(&self) -> Duration {
```
Comment (LucioFranco) [inline src/consensus/mod.rs:183]: could put this in `src/common.rs`

**Review by LucioFranco — APPROVED** [review]
Looks really good left some comments inline.

File: src/consensus/mod.rs (hunk 2 above)
Comment (bIgBV) [reply src/consensus/mod.rs:79]: Yeah you are right, it is. So how _would_ you schedule a background task?

File: src/consensus/mod.rs (hunk 10 above)
Comment (bIgBV) [reply src/consensus/mod.rs:183]: will it actually be used anywhere outside of this module though?

Comment (LucioFranco) [reply src/consensus/mod.rs:183]: maybe 🤷‍♂

**Pull request merged**